Epic 9975 : Adding support for copying QC Folders

labkey-bpatel opened the pull request:
#### Rationale
In the interest of publishing QC Folders to Panorama public, handle exporting and importing of QC properties such as: 
New custom metrics and trace metrics, Metric configurations and settings, Guide sets, Excluded samples, Excluded Precursors, Annotations, and Plot settings from default view.

#### Related Pull Requests
https://github.com/LabKey/platform/pull/3229

#### Changes
* Framework to handle folder export and import
* Export Plot settings from default view to .properties file
* Export metrics, exclusions, guidesets, annotations to .tsv file only if data is present
* Handle exporting of "name" instead of Id
* Save Plot settings from .properties file
* Import metrics, exclusions, guidesets, annotations from .tsv to appropriate tables
* Resolve "name" into Id during import
* Skip inserting row if the data is already present

labkey-bpatel committed:
Framework to import/export QC metrics during folder import/export.

labkey-bpatel committed:
Update comments

labkey-bpatel committed:
Update data type name and dir name.

labkey-bpatel committed:
Show 'QC settings' in folder export for QC folders only.

labkey-bpatel committed:
Export to .tsv for Metrics, Guidesets, Exclusions, and Annotations tables

labkey-bpatel committed:
Merge remote-tracking branch 'origin/release22.3-SNAPSHOT' into 22.3_fb_copy_qc_folders

labkey-bpatel committed:
Handle exporting QC Plot settings to .properties file during folder export. Add 'excludedPrecursors' to LeveyJenningsPlotOptions.

labkey-bpatel committed:
Refactor export code/cleanup

labkey-bpatel committed:
Import qc plot settings.

labkey-bpatel committed:
Implement tsv import (wip)

labkey-bpatel committed:
Change QCEnabledMetricsTable to extend SimpleTable to be able to bulk import/make it UpdatableTableInfo

labkey-bpatel committed:
Resolve rowId/name for QCEnabledMetrics

labkey-bpatel committed:
Merge remote-tracking branch 'origin/release22.3-SNAPSHOT' into 22.3_fb_copy_qc_folders

labkey-bpatel committed:
Update export code to include Name instead of Id. Handle import of QCAnnotation and QCMetricExclusion. Make QCAnnotationTypeTable extend SimpleTable.

labkey-bpatel committed:
Handle import of Annotations, Guidesets, Exclusions. Handle duplicates, code refactor and cleanup.

labkey-bpatel committed:
Minor refactor

labkey-bpatel committed:
Minor refactor

labkey-bpatel committed:
Handle replicateAnnotation export with name instead of id. Refactor common code. Update logging to be more readable.

labkey-bpatel committed:
Merge remote-tracking branch 'origin/release22.3-SNAPSHOT' into 22.3_fb_copy_qc_folders

labkey-bpatel committed:
Log msg update

labkey-jeckels commented on src/org/labkey/targetedms/folderImport/PanoramaQCSettings.java:
```diff
@@ -0,0 +1,478 @@
+                    }
+                    catch (Exception e)
+                    {
+                        LOG.error("Error exporting 'default view QC Plot settings'", e);
```
Are you expecting errors here? Rethrowing is probably better to make sure we ultimately learn of them and adjust as needed

labkey-jeckels commented on src/org/labkey/targetedms/folderImport/PanoramaQCSettings.java:
```diff
@@ -0,0 +1,478 @@
+                }
+                catch (Exception e)
+                {
+                    LOG.error("Error writing results to " + fileName, e);
```
Same here - recommend rethrowing

labkey-jeckels commented on src/org/labkey/targetedms/folderImport/QCFolderImporter.java:
```diff
@@ -0,0 +1,110 @@
+               {
+                   if (qcSetting.getSettingsFileName().equalsIgnoreCase(QCFolderConstants.QC_PLOT_SETTINGS_PROPS_FILE_NAME))
+                   {
+                       LOG.error("Error importing QC Plot settings from " + QCFolderConstants.QC_PLOT_SETTINGS_PROPS_FILE_NAME + ": " + e.getMessage(), e);
```
Like with the other errors, this means that nothing will get written to the job's log file and the import will be reported as successful.

labkey-bpatel replied:
good point, rethrowing exception now

labkey-jeckels reviewed (commented):
From a quick scan, this all looks reasonable. A couple of suggestions around exception handling. I can take a closer look when dev is complete.

labkey-bpatel committed:
Throw exception instead of just logging error. Minor code refactor.

labkey-jeckels commented on src/org/labkey/targetedms/folderImport/PanoramaQCSettings.java:
```diff
@@ -0,0 +1,489 @@
+                {
+                    TargetedMSSchema schema = new TargetedMSSchema(user, container);
+                    TableInfo ti = schema.getTable(getTableName());
+                    assert ti != null;
```
I always find this style of assert useless, though I know we have it scattered throughout our code. On a dev machine with assertions turned on, it will fail with an assert on this line, with no message. On production servers it'll fail with a NPE on the next line. What's the advantage to asserting?

I'm guessing this was added because IntelliJ complained that `ti` might be null. @labkey-matthewb maybe we should have a UserSchema.getTableOrFail() method (or whatever we call it) that throws an IllegalArgumentException if the schema doesn't have the table. We can annotate that method as NotNull and have it fail with an error that includes the table name that failed to resolve.

labkey-jeckels commented on src/org/labkey/targetedms/folderImport/QCFolderImporter.java:
```diff
@@ -0,0 +1,106 @@
+               {
+                   if (qcSetting.getSettingsFileName().equalsIgnoreCase(QCFolderConstants.QC_PLOT_SETTINGS_PROPS_FILE_NAME))
+                   {
+                       throw new Exception("Error importing QC Plot settings from " + QCFolderConstants.QC_PLOT_SETTINGS_PROPS_FILE_NAME + ": " + e.getMessage(), e);
```
Throwing an ImportException instead of the base class is likely a bit friendlier

labkey-jeckels commented on src/org/labkey/targetedms/folderImport/QCFolderConstants.java:
```diff
@@ -0,0 +1,21 @@
+{
+    protected static final String QC_FOLDER_DATA_TYPE = "Panorama QC Folder Settings";
+    protected static final String QC_FOLDER_DIR = "PanoramaQC";
+    protected static final String CATEGORY = "TargetedMSLeveyJenningsPlotOptions";
```
Can we get rid of the separate constant in TargetedMSController now?

labkey-jeckels commented on src/org/labkey/targetedms/folderImport/PanoramaQCSettings.java:
```diff
@@ -0,0 +1,489 @@
+        if (new TableSelector(ti, row.keySet(), filter, null).getRowCount() > 0)
+        {
+            logMsg += "] values already exists. Skipping";
+            ctx.getLogger().warn(logMsg);
```
This feels like info, not warning

labkey-jeckels commented on src/org/labkey/targetedms/folderImport/PanoramaQCSettings.java:
```diff
@@ -0,0 +1,489 @@
+                }
+                catch (IOException e)
+                {
+                    throw new IOException("Error writing results to " + fileName, e);
```
Can this be consolidated? Seems overkill to catch IOException, throw a new IOException, and then catch that all within 5 lines

labkey-jeckels commented on src/org/labkey/targetedms/folderImport/PanoramaQCSettings.java:
```diff
@@ -0,0 +1,489 @@
+                    tableMap.put("qcAnnotation", ti);
+                    tableMap.put("qcAnnotationType", schema.getTable(TargetedMSSchema.TABLE_QC_ANNOTATION_TYPE));
+
+                    ResultsFactory factory = ()-> QueryService.get().selectResults(schema, sql.getSQL(), tableMap, null, true, true);
```
Why do we need to pass in the tableMap here? Can't QueryService resolve the tables itself from the schema?

labkey-bpatel committed:
Code review changes:
- Remove constant 'CATEGORY' from TargetedMSController, use from QCFolderConstants
- Throw ImportException instead of Exception
- Remove assert statments, use null check and throw exception instead
- Remove usage of tableMap for SQLFragment
- Other minor refactors related to renaming of methods, and improving log messages

labkey-jeckels commented on src/org/labkey/targetedms/folderImport/PanoramaQCSettings.java:
```diff
@@ -330,17 +310,37 @@ public String getSettingsFileName()
+        TargetedMSSchema schema = new TargetedMSSchema(user, container);
+        TableInfo ti = schema.getTable(tableName, cf);
+        if (null == ti)
+            throw new NotFoundException(schema.getSchemaName() + "." + getTableName() + " not found in '" + container.getName() + "'");
```
container.getPath() is a bit more helpful

labkey-bpatel committed:
Handle null values.

labkey-bpatel committed:
Bug fixing:
- Handle annotations from the Shared folder
- Handle case when there is a mismatch between a Text column and numeric value.

labkey-bpatel committed:
Close results to avoid 'CachedResultSet was not closed' error

labkey-adam commented on src/org/labkey/targetedms/folderImport/PanoramaQCSettings.java:
```diff
@@ -53,8 +54,10 @@ public void exportSettings(VirtualFile vf, Container container, User user) throw
+                    Results results = QueryService.get().select(ti, userEditableCols, filter, null);
+                    ResultsFactory factory = ()-> results;
                     exportSettingsToTSV(vf, factory, getSettingsFileName(), getTableName());
+                    results.close();
```
This and the six other similar changes in this commit don't look right to me. The whole point of passing a factory into `exportSettingsToTSV()` (which then passes it into `TSVGridWriter`) is to allow the calling code to determine how a `Results` is generated, but to give the responsibility for actually generating and closing the `Results` to the consuming code (TSVGridWriter, in this case). Passing in a factory allows the consumer to use try-with-resources to handle the results (much preferred over passing around open Results objects). This change is not going to fix any "CachedResultSet not closed" errors... invoking the factory and closing that Results here just executes an extra query and throws away the results; it doesn't in any way affect the Results generated by the downstream code after it re-invokes the factory.

Now I see the problem... `exportSettingsToTSV()` invokes `factory.get().countAll()`... this call to the factory generates the CachedResultSet that's never closed. I'm sure you want to execute the query only once (current code is executing it three times), so generate and stash a Results, check countAll(), and pass a new factory `()->results` to TSVGridWriter. You have to close the results in the count == 0 case, so might as well generate the Results in a try-with-resources. This means it will be double closed in most cases, but CachedResultSet doesn't care.

labkey-adam replied:
This is difficult to explain in a commit comment... happy to discuss, just let me know

labkey-bpatel replied:
This change did fix the "CachedResultSet not closed" errors: 
https://teamcity.labkey.org/viewType.html?buildTypeId=LabKey_223Release_Premium_ModulesSuites_PanoramaPremiumPostgres&branch_LabKey_223Release_Premium_ModulesSuites=copy_qc_folders&tab=buildTypeStatusDiv

Sounds good, lets sync on Monday! Thanks!

labkey-adam replied:
We discussed today, but for the record: I missed that invoking the factory doesn't execute the query... so it makes sense that this change fixed the exception and doesn't actually execute the query multiple times (as I had claimed). I still think it would be cleaner for each caller to simply pass in a query-executing factory and for `exportSettingsToTSV()` to invoke the factory in a try-with-resources, check countAll(), and pass on a `()->results` factory if appropriate.

labkey-bpatel replied:
Hi Adam - just seeing your followup comment, thanks for clarifying and insights. I've made some updates, please review. Thank you!

labkey-jeckels commented on src/org/labkey/targetedms/folderImport/PanoramaQCSettings.java:
```diff
@@ -0,0 +1,533 @@
+                @Override
+                public int importSettingsFromFile(FolderImportContext ctx, VirtualFile panoramaQCDir, @Nullable TargetedMSSchema schema, @Nullable TableInfo ti, @Nullable QueryUpdateService qus, @Nullable BatchValidationException errors) throws Exception
+                {
+                    DataLoader loader = new TabLoader(Readers.getReader(panoramaQCDir.getInputStream(getSettingsFileName())), true);
```
This and other TabLoader instances need to be closed, ideally in a try-with-resources block.

labkey-bpatel replied:
Done! Thanks, Josh!

labkey-bpatel committed:
Use try-wtih-resources to close out Results and TabLoader instances

labkey-bpatel committed:
Minor update to try-with-resources statement

labkey-bpatel committed:
Pass ()-> results to the TSVGridWriter

labkey-adam commented:
I was suggesting you pass in a factory to `exportSettingsToTSV()`. You're generating and passing in a `Results` instead... that's fine, actually, it's cleaner than my suggestion. But the code generating the Results should (almost) always close it in a try-with-resources. See inline comments.

labkey-adam commented on src/org/labkey/targetedms/folderImport/PanoramaQCSettings.java:
```diff
@@ -0,0 +1,524 @@
+                    List<ColumnInfo> userEditableCols = ti.getColumns().stream().filter(ci -> ci.isUserEditable()).collect(Collectors.toList());
+                    SimpleFilter filter = SimpleFilter.createContainerFilter(container); //only export the ones that are defined in current container (and not the ones from the root container)
+
+                    Results results = QueryService.get().select(ti, userEditableCols, filter, null);
```
try-with-resources

labkey-adam commented on src/org/labkey/targetedms/folderImport/PanoramaQCSettings.java:
```diff
@@ -0,0 +1,524 @@
+                            .append(TargetedMSManager.getTableInfoQCMetricConfiguration(), "qcMetricConfig")
+                            .append(" ON qcEnabledMetrics.metric = qcMetricConfig.Id");
+
+                    Results results = QueryService.get().selectResults(schema, sql.getSQL(), null, null, true, true);
```
try-with-resources

labkey-adam commented on src/org/labkey/targetedms/folderImport/PanoramaQCSettings.java:
```diff
@@ -0,0 +1,524 @@
+                    TableInfo ti = getTableInfo(user, container, cf);
+                    List<ColumnInfo> userEditableCols = ti.getColumns().stream().filter(ci -> ci.isUserEditable()).collect(Collectors.toList());
+
+                    Results results = QueryService.get().select(ti, userEditableCols, null, null);
```
try-with-resources

labkey-adam commented on src/org/labkey/targetedms/folderImport/PanoramaQCSettings.java:
```diff
@@ -0,0 +1,524 @@
+
+                    Results results = QueryService.get().select(ti, userEditableCols, null, null);
+                    exportSettingsToTSV(vf, results, getSettingsFileName(), getTableName());
+                    results.close();
```
Remove... always use try-with-resources instead of explicitly calling close() (unless there's a really good reason)

labkey-adam commented on src/org/labkey/targetedms/folderImport/PanoramaQCSettings.java:
```diff
@@ -0,0 +1,524 @@
+                            .append(" SELECT * FROM \"Shared\".targetedms.qcannotationtype) qcAnnotationType")
+                            .append(" ON qcAnnotation.QCAnnotationTypeId = qcAnnotationType.Id");
+
+                    Results results = QueryService.get().selectResults(schema, sql.getSQL(), null, null, true, true);
```
try-with-resources

labkey-adam commented on src/org/labkey/targetedms/folderImport/PanoramaQCSettings.java:
```diff
@@ -0,0 +1,524 @@
+                            .append(" ON replicate.RunId = runs.Id")
+                            .append(" WHERE Source != 'Skyline' ");
+
+                    Results results = QueryService.get().selectResults(schema, sql.getSQL(), null, null, true, true);
```
try-with-resources

labkey-adam commented on src/org/labkey/targetedms/folderImport/PanoramaQCSettings.java:
```diff
@@ -0,0 +1,524 @@
+                            .append(TargetedMSManager.getTableInfoRuns(), "runs")
+                            .append(" ON replicate.RunId = runs.Id");
+
+                    Results results = QueryService.get().selectResults(schema, sql.getSQL(), null, null, true, true);
```
try-with-resources

labkey-adam commented on src/org/labkey/targetedms/folderImport/PanoramaQCSettings.java:
```diff
@@ -0,0 +1,524 @@
+        TableInfo ti = getTableInfo(user, container, null);
+        List<ColumnInfo> userEditableCols = ti.getColumns().stream().filter(ci -> ci.isUserEditable()).collect(Collectors.toList());
+
+        Results results = QueryService.get().select(ti, userEditableCols, null, null);
```
try-with-resources

labkey-adam commented on src/org/labkey/targetedms/folderImport/PanoramaQCSettings.java:
```diff
@@ -0,0 +1,524 @@
+    protected void exportSettingsToTSV(VirtualFile vf, Results results, String fileName, String tableName) throws Exception
+    {
+        //try-with-resources to ensure Results instance is closed regardless of whether try statement executes as expected or throws an exception
+        try (results)
```
Remote this try since callers are generating the `Results` (and therefore responsible for closing it)

labkey-bpatel replied:
Good point! done

labkey-bpatel committed:
Move try-with-resources in the caller. Add BufferedReader to try.

labkey-jeckels commented on src/org/labkey/targetedms/folderImport/PanoramaQCSettings.java:
```diff
@@ -0,0 +1,535 @@
+                    List<Map<String, Object>> tsvData = getTsvData(panoramaQCDir, getSettingsFileName());
+                    List<Map<String, Object>> dataWithoutDuplicates = new ArrayList<>();
+
+                        tsvData.forEach(row -> {
+                            String metricValue = (String) row.get("metric");
+                            Integer metricId = getRowIdFromName(metricValue, getSettingsFileName(), getTableName(), TargetedMSSchema.TABLE_QC_METRIC_CONFIGURATION, Set.of("Id"), new SimpleFilter(FieldKey.fromParts("Name"), metricValue), ctx.getUser(), ctx.getContainer(), null);
+                            row.put("metric", metricId);
+                            getDataWithoutDuplicates(ctx, ti, row, dataWithoutDuplicates);
+                        });
```
This is indented an extra level

labkey-jeckels commented on src/org/labkey/targetedms/folderImport/PanoramaQCSettings.java:
```diff
@@ -0,0 +1,535 @@
+                {
+                    tsvWriter.setApplyFormats(false);
+                    tsvWriter.setColumnHeaderType(ColumnHeaderType.FieldKey);
+                    PrintWriter out = vf.getPrintWriter(fileName);
```
The PrintWriter should be closed via a try-with-resources as well

labkey-jeckels commented on src/org/labkey/targetedms/folderImport/PanoramaQCSettings.java:
```diff
@@ -0,0 +1,535 @@
+            }
+            else
+            {
+                throw new NotFoundException("Query update service for targetedms." + tableName + " not found.");
```
NotFoundException is intended for HTTP requests where the client sent a RowId or other identifier that the server couldn't find. IllegalArgumentException would be more appropriate here.

labkey-jeckels commented on src/org/labkey/targetedms/folderImport/PanoramaQCSettings.java:
```diff
@@ -0,0 +1,535 @@
+                }
+                catch (DuplicateKeyException | BatchValidationException | QueryUpdateServiceException | SQLException e)
+                {
+                    throw new Exception("Data from " + settingsFileName + " did not get imported into targetedms." + tableName, e);
```
ImportException is a more specific subclass that seems appropriate here.

labkey-jeckels commented on src/org/labkey/targetedms/folderImport/PanoramaQCSettings.java:
```diff
@@ -0,0 +1,535 @@
+            else
+                msg += "Unable to import data from " + fileName + " into targetedms." + targetTable;
+
+            throw new NotFoundException(msg);
```
See comment below about NotFoundException vs IllegalArgumentException

labkey-jeckels commented on src/org/labkey/targetedms/folderImport/PanoramaQCSettings.java:
```diff
@@ -0,0 +1,535 @@
+        }
+        catch (IOException | SQLException e)
+        {
+            throw new Exception("Error getting results from " + tableName, e);
```
See other comment about ImportException vs Exception. There are a few other examples

labkey-jeckels reviewed (changes requested):
A few final cleanup items

labkey-adam reviewed (approved):
Calls to `exportSettingsToTSV()` look good to me. Josh is reviewing everything else.

labkey-bpatel committed:
Code review changes:
Use IllegalArgumentException or ImportException.
Add PrintWriter to try
